Facing and Axis to enum

DebHugo opened the pull request:
Fix #89

This change requires many changes in pmmp. I can do if this pr is good for you.

DebHugo committed:
Facing and Axis to enum

DebHugo committed:
Modif toString function

dktapps commented on src/Facing.php:
```diff
@@ -56,6 +56,9 @@ private function __construct(){
+	/**
+	 * @deprecated
+	 */
 	public const OFFSET = [
```
why is this marked as deprecated?

DebHugo replied:
It's marked as obsolete because i'm replacing it with a function. In my opinion, it is clearer.

With constant:
```
Facing::OFFSET[Facing::DOWN->value]
```

With function:
```
Facing::DOWN->offset()
```

dktapps replied:
This seems outside of the PR scope

dktapps commented on src/Facing.php:
```diff
@@ -89,67 +95,60 @@ private function __construct(){
 	 * Returns the opposite Facing of the specified one.
 	 *
-	 * @param int $direction 0-5 one of the Facing::* constants
+	 * @throws \ValueError if opposite facing don't exist
```
This is impossible

dktapps commented:
If we're going with enums I would prefer to ditch the numeric values entirely. They have no purpose except to be implicitly misused as Minecraft constants.

DebHugo committed:
Remove useless doc

DebHugo committed:
remove outside functionality of this pr

DebHugo committed:
Ditch numeric values

DebHugo commented:
> If we're going with enums I would prefer to ditch the numeric values entirely. They have no purpose except to be implicitly misused as Minecraft constants.

Should I do the same for Axis? I don't know if these values ​​represent anything or not

dktapps commented:
> > If we're going with enums I would prefer to ditch the numeric values entirely. They have no purpose except to be implicitly misused as Minecraft constants.
> 
> Should I do the same for Axis? I don't know if these values ​​represent anything or not

Yes

DebHugo committed:
Remove ditch values

dktapps commented on src/Facing.php:
```diff
@@ -89,67 +88,58 @@ private function __construct(){
+	public static function opposite(Facing $direction) : Facing{
+		return match ($direction) {
+			self::DOWN => self::UP,
+			self::EAST => self::SOUTH,
```
This is incorrect

dktapps commented on src/Facing.php:
```diff
@@ -89,67 +88,58 @@ private function __construct(){
+			self::EAST => self::SOUTH,
+			self::WEST => self::EAST,
+			self::UP => self::DOWN,
+			self::SOUTH => self::EAST,
```
also incorrect

DebHugo committed:
Fix incorrect opposite side

DebHugo committed:
Merge branch 'enum' of https://github.com/Dhaiven/Math into enum

dktapps commented on src/Facing.php:
```diff
@@ -89,67 +88,58 @@ private function __construct(){
-	public static function isPositive(int $direction) : bool{
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public static function isPositive(Facing $direction) : bool{
+		return in_array($direction, [self::UP, self::SOUTH, self::EAST]);
```
`in_array()` is an immediate red flag whenever I see it

dktapps commented on src/Facing.php:
```diff
@@ -89,67 +88,58 @@ private function __construct(){
+			self::WEST => self::EAST,
+			self::UP => self::DOWN,
+			self::SOUTH => self::NORTH,
+			self::EAST => self::WEST,
```
This should follow the order of the cases - it's a confusing mess right now

Dhaiven replied:
How i can do ?

dktapps commented on src/Facing.php:
```diff
@@ -89,67 +88,58 @@ private function __construct(){
-
-		$rotated = self::CLOCKWISE[$axis][$direction];
+	public static function rotate(Facing $direction, Axis $axis, bool $clockwise) : Facing{
+		$rotated = self::CLOCKWISE[$axis->value][$direction->value];
```
These don't have values anymore

dktapps commented on src/Facing.php:
```diff
@@ -159,17 +149,17 @@ public static function validate(int $facing) : void{
 	}
```
There's no point deprecating this. We're making sweeping BC breaks here anyway.

DebHugo replied:
So i remove this function ?

dktapps commented on src/Facing.php:
```diff
@@ -159,17 +149,17 @@ public static function validate(int $facing) : void{
 	}
 
 	/**
+	 * @deprecated use Facing->name
```
If this is a valid replacement, why isn't the function using it?

DebHugo replied:
Indead

DebHugo committed:
Modifications

DebHugo committed:
Follow order of cases

DebHugo committed:
Fix bug in rotate function

DebHugo committed:
Modification

DebHugo commented on src/AxisAlignedBB.php:
```diff
@@ -136,20 +136,18 @@ public function offsetCopy(float $x, float $y, float $z) : AxisAlignedBB{
-	public function offsetTowards(int $face, float $distance) : AxisAlignedBB{
-		[$offsetX, $offsetY, $offsetZ] = Facing::OFFSET[$face] ?? throw new \InvalidArgumentException("Invalid Facing $face");
+	public function offsetTowards(Facing $face, float $distance) : AxisAlignedBB{
+		[$offsetX, $offsetY, $offsetZ] = Facing::OFFSET[$face->value];
```
Since faces don't have values, this creates a bug. To fix this bug, why not introduce an offset() function like this:
```
public function offset(): array {
		return match($this){
			self::DOWN  => [ 0, -1,  0],
			self::UP    => [ 0, +1,  0],
			self::NORTH => [ 0,  0, -1],
			self::SOUTH => [ 0,  0, +1],
			self::WEST  => [-1,  0,  0],
			self::EAST  => [+1,  0,  0]
		};
	}

#Result
Face::DOWN->offset()
```

I already tried to add this previously and you say it's outside of this pr.
I agree but I don't see any other solutions to solve this problem.
I await your opinion.

dktapps replied:
Yeah, I don't see another way to do it without bringing back the values.

dktapps committed:
Merge branch 'stable' into enum

DebHugo committed:
Updates

DebHugo committed:
Merge branch 'enum' of https://github.com/Dhaiven/Math into enum

DebHugo committed:
Fix bugs

DebHugo committed:
Fix cs

DebHugo committed:
Fix cs

DebHugo commented:
This pr is good ? if yes i can change pocketmine to support Facing enum

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
 	 */
-	public static function axis(int $direction) : int{
-		return $direction >> 1; //shift off positive/negative bit
+	public static function axis(Facing $direction) : Axis{
```
Doesn't make sense for this to still be static

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
 	 */
-	public static function isPositive(int $direction) : bool{
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public static function isPositive(Facing $direction) : bool{
```
As above

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
 	 */
-	public static function opposite(int $direction) : int{
-		return $direction ^ self::FLAG_AXIS_POSITIVE;
+	public static function opposite(Facing $direction) : Facing{
```
As above

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
-		if(!isset(self::CLOCKWISE[$axis][$direction])){
-			throw new \InvalidArgumentException("Cannot rotate facing \"" . self::toString($direction) . "\" around axis \"" . Axis::toString($axis) . "\"");
-		}
+	public static function rotate(Facing $direction, Axis $axis, bool $clockwise) : Facing{
```
As above

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
-	 * @throws \InvalidArgumentException
-	 */
-	public static function rotateY(int $direction, bool $clockwise) : int{
+	public static function rotateY(Facing $direction, bool $clockwise) : Facing{
```
As above

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
-	 * @throws \InvalidArgumentException
-	 */
-	public static function rotateZ(int $direction, bool $clockwise) : int{
+	public static function rotateZ(Facing $direction, bool $clockwise) : Facing{
```
As above

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
-	 * @throws \InvalidArgumentException
-	 */
-	public static function rotateX(int $direction, bool $clockwise) : int{
+	public static function rotateX(Facing $direction, bool $clockwise) : Facing{
```
As above

dktapps commented on src/Vector3.php:
```diff
@@ -170,19 +170,19 @@ public function east(int $step = 1){
-		foreach(Facing::ALL as $facing){
-			yield $facing => $this->getSide($facing, $step);
+		foreach(Facing::cases() as $facing){
+			yield [$facing, $this->getSide($facing, $step)];
```
This didn't need to be changed. Generators allow yielding objects as the key.

dktapps commented on src/Vector3.php:
```diff
@@ -170,19 +170,19 @@ public function east(int $step = 1){
 	 * Same as sides() but returns a pre-populated array instead of Generator.
 	 *
-	 * @return Vector3[]
+	 * @return array<int, array{Facing,Vector3}>
```
Same here

dktapps commented on src/Vector3.php:
```diff
@@ -191,15 +191,13 @@ public function sidesArray(bool $keys = false, int $step = 1) : array{
+		foreach(Facing::cases() as $facing){
 			if(Facing::axis($facing) !== $axis){
-				yield $facing => $this->getSide($facing, $step);
+				yield [$facing, $this->getSide($facing, $step)];
```
Same here

DebHugo committed:
Replace static functions by object functions

DebHugo committed:
Add facing to key for Generator

DebHugo committed:
Fix tests

DebHugo commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
 		self::EAST
 	];
```
To keep the new logic, i think `HORIZONTAL` constant must be removed and replace by 2 functions:
- `horizontal(): array` -> return same value likes `HORIZONTAL` constant
- `isHorizontal(): bool` -> return true if the face is horizontal else false
I await your opinion

dktapps replied:
I don't think that's necessary. This is only used for iteration anyway.

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,92 @@ private function __construct(){
 
 	/**
-	 * Returns whether the direction is facing the positive of its axis.
+	 * @return non-empty-array<int>
```
PHPStan doc comments should be prefixed by @phpstan-

DebHugo committed:
Prefix phpstan return

DebHugo committed:
Update documentation

dktapps commented on src/AxisAlignedBB.php:
```diff
@@ -465,20 +448,23 @@ public function calculateIntercept(Vector3 $pos1, Vector3 $pos2) : ?RayTraceResu
 				$vector = $v;
 				$distance = $d;
-				$face = $f;
+				/** @var Facing $face */
```
Don't use @var tags like this, they'll suppress legit errors

dktapps commented on src/AxisAlignedBB.php:
```diff
@@ -465,20 +448,23 @@ public function calculateIntercept(Vector3 $pos1, Vector3 $pos2) : ?RayTraceResu
 		$vector = null;
 		$distance = PHP_INT_MAX;
-		$face = -1;
+		/** @var Facing $face */
```
Same here

DebHugo committed:
Remove @var to specific content of variable

dktapps committed:
Update AxisAlignedBB.php

dktapps committed:
smh

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
+	}
+
+	/**
+	 * @phpstan-return non-empty-array<int>
```
```suggestion
	 * @phpstan-return array{-1|0|1, -1|0|1, -1|0|1}
```

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
+				self::EAST => self::SOUTH,
+				self::SOUTH => self::WEST,
+				self::WEST => self::NORTH,
+				default => throw new \InvalidArgumentException("Face " . strtolower($this->name) . " not match with Axis " . strtolower($axis->name))
```
These error messages should match the originals

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
+				self::EAST => self::DOWN,
+				self::DOWN => self::WEST,
+				self::WEST => self::UP,
+				default => throw new \InvalidArgumentException("Face " . strtolower($this->name) . " not match with Axis " . strtolower($axis->name))
```
and here

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
+				self::NORTH => self::DOWN,
+				self::DOWN => self::SOUTH,
+				self::SOUTH => self::UP,
+				default => throw new \InvalidArgumentException("Face " . strtolower($this->name) . " not match with Axis " . strtolower($axis->name))
```
and here

dktapps commented on tests/phpunit/FacingTest.php:
```diff
@@ -35,9 +35,9 @@ public static function axisProvider() : \Generator{
-	#[DataProvider("axisProvider")]
-	public function testAxis(int $direction, int $axis) : void{
-		self::assertEquals($axis, Facing::axis($direction));
+  #[DataProvider("axisProvider")]
```
Unrelated formatting change

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
 	/**
+	 * Rotates the given direction around the X axis.
+	 * 
+	 * @see Facing::rotate()
```
Unrelated change

DebHugo replied:
I keep the @throws \InvalidArgumentException or not ?

dktapps replied:
Just remove the stuff that was added that isn't related to the PR

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
 	/**
+	 * Rotates the given direction around the Z axis.
+	 * 
+	 * @see Facing::rotate()
```
Unrelated

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
 	/**
+	 * Rotates the given direction around the Y axis.
+	 * 
+	 * @see Facing::rotate()
```
Unrelated

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
-	public static function isPositive(int $direction) : bool{
-		return ($direction & self::FLAG_AXIS_POSITIVE) === self::FLAG_AXIS_POSITIVE;
+	public function isPositive() : bool{
+		return $this === self::UP || $this === self::SOUTH || $this === self::EAST;
```
I'd prefer if this used a match with all cases.
It doesn't affect anything in this particular case, but it's generally better hygiene for enums, since it'll guarantee an error will appear if more cases are added.

DebHugo replied:
Like this
```
return match($this){
    self::UP, self::SOUTH, self::EAST => true,
    self::DOWN, self::NORTH, self::WEST => false,
    default => throw new \InvalidArgumentException("Cannot know if facing \"" . strtolower($this->name) . "\"is positive or not")
};
```
or witth `default => false` ?

dktapps replied:
Like that, but delete the `default` case

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
-	public static function axis(int $direction) : int{
-		return $direction >> 1; //shift off positive/negative bit
+	public function axis() : Axis{
+		return match ($this) {
```
```suggestion
		return match($this){
```

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
+	 * @phpstan-return non-empty-array<int>
+	 */
+	public function offset() : array{
+		return match ($this) {
```
```suggestion
		return match($this){
```

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
-	public static function opposite(int $direction) : int{
-		return $direction ^ self::FLAG_AXIS_POSITIVE;
+	public function opposite() : Facing{
+		return match ($this) {
```
```suggestion
		return match($this){
```

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
 	 * Rotates the given direction around the axis.
 	 *
-	 * @throws \InvalidArgumentException if not possible to rotate $direction around $axis
+	 * @throws \InvalidArgumentException
```
Unrelated change

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
-			throw new \InvalidArgumentException("Cannot rotate facing \"" . self::toString($direction) . "\" around axis \"" . Axis::toString($axis) . "\"");
-		}
+	public function rotate(Axis $axis, bool $clockwise) : Facing{
+		$rotated = match ($axis) {
```
```suggestion
		$rotated = match($axis){
```

dktapps committed:
formatting tidy

I was going to submit these as review comments, but it's less effort if i just fix them directly

DebHugo committed:
Remove unrealated changes

DebHugo committed:
Cs

dktapps commented on src/Facing.php:
```diff
@@ -56,120 +38,112 @@ private function __construct(){
+				self::DOWN => self::SOUTH,
+				self::SOUTH => self::UP,
+				default => throw new \InvalidArgumentException("Cannot rotate facing \"" . strtolower($this->name) . "\" around axis \"" . strtolower($axis->name) . "\"")
+			}
```
The default cases could just fall through to `null`, then we could just check if `$rotated` is `null` after the match so as not to duplicate exception messages 3 times

DebHugo committed:
Remove unrelated documentation

DebHugo committed:
Don't repeat error or condition

dktapps committed:
Update Facing.php

dktapps committed:
Update Vector3.php

dktapps committed:
github web editor try not to suck challenge IMPOSSIBLE

dktapps committed:
Update Facing.php
